### ISSUE TITLE
.once not triggering consistently when Gun is used inside a ServiceWork

### ISSUE THREAD

**Issue by davide** [amark/gun#681]
I've created a library called [p2p-fetch](https://github.com/davide/p2p-fetch) that under the hood uses Gun (currently version 0.9.999995).

The library employs a ServiceWorker to catch web requests in flight and if they happen to match a configured FETCH_PATTERN instead of loading the resource from the web a lookup is done to see if that asset is already cached in [distributed] Gun database. If it is already there then the Gun stored content is used, otherwise it fallbacks to a regular web request (storing the content of the web response in the Gun database for reuse by others).

To stress test the Gun logic  I included an additional parameter FORCE_P2P which will pass in to the get(...).once(...) call a big timeout value to make sure that it never fallbacks to the regular web request.

During those tests I started noticing that *sometimes* the test page would have requests hanging, never triggering the .once callback. For the p2p-fetch library having the .once callback called is critical, even if it's to inform that there's no data to start the web fallback. 

After some debugging it seems that all those failed requests go through this line:
https://github.com/amark/gun/blob/9f9b3e557c1b9f9ff3d774edcec30271a596d19a/src/chain.js#L91
Which funny enough has a TODO and a reference to a possible bug. 😄 

Here are the steps to replicate the issue:
 - Open this page https://lego-star-wars-fiwonixfri.now.sh/ on Google chrome to get the assets loaded;
 - On the same browser then open this URL: https://lego-star-wars-fiwonixfri.now.sh/?FORCE_P2P=true
After a refresh some of the 3 images might not show up.

The source code for that example page lives here:
https://github.com/davide/p2p-fetch/tree/master/examples/lego-star-wars

I haven't excluded the possibility of a messed up interaction / lack of readiness from the ServiceWorker but wanted to report this issue so that anyone more experienced with the Gun API can take a peek.

Thanks in advance for any help or tips on how to further debug this issue.

**Comment by davide** [comment]
Removed [some code that was calling .once inside another .once callback](https://github.com/davide/p2p-fetch/commit/5ae5faa9b3354f4ce73df79e6e098a343526456d).

New demo with that new code:
https://lego-star-wars-njfrsbrzyd.now.sh/
https://lego-star-wars-njfrsbrzyd.now.sh/?FORCE_P2P=true

Noticed that during the first page load *after closing and opening the browser* no .once callbacks are triggered, but waiting around 30s and then refreshing the page they are serving the data from the Gun database.

**Comment by davide** [comment]
Here's what you should see on the console during that first load (it won't budge from this):
![outsystems success journey](https://user-images.githubusercontent.com/40282/50875727-90c5aa80-13c1-11e9-8f96-d36cefc2a441.png)

**Comment by davide** [comment]
Here's what you should see during that first load (it won't budge from this):
![p2p-fetch_first_load](https://user-images.githubusercontent.com/40282/50875838-0c275c00-13c2-11e9-8eec-a63b149b43d2.png)

**Comment by bugs181** [comment]
Hey, sorry you're running into this bug! Just curious if you could try to do the logic in the callback of the `gun.put` instead? That may solve some issues.

```
gun.get(encodeURI(url)).put(..., function(ack) {
                   resolve(
	                  new Response(blob, init)
	                );
})
```

**Comment by davide** [comment]
Hi Levi! Just applied that patch but no changes:
![p2p-fetch_first_load-v2](https://user-images.githubusercontent.com/40282/50876109-64ab2900-13c3-11e9-82be-2da31b310e0c.png)

**Comment by davide** [comment]
After a bit the 3 requests fail with this error: "Failed to load resource: net::ERR_FAILED"

**Comment by bugs181** [comment]
I don't seem to be receiving any console messages about Gun loading on Safari 12.

**Comment by amark** [comment]
Hmm, I wonder if this also relates to an issue @mmalmi found... which I think his might have to do with `.once` getting the metadata / empty object and then later (but not within `.once` debounce's timeout) getting the rest of the data. This note is mostly just a reminder of myself of this... still need to dig into this thread more - was dealing with the security upgrades earlier this week and just now starting to have more time, but still trying to process things in various priority orders, so I apologize for the delay!

**Comment by davide** [comment]
Sorry for the delay @bugs181. RL...

When using service workers there are multiple parallel threads in place (1 for the ServiceWorker and 1 for each tab). The first tab is the one that kicks off the ServiceWorker thread - which is where Gun is loaded (only once).
Loading a new tab doesn't get Gun to load again.
If you see the "P2P FETCH: ServiceWorker active" message then the ServiceWorker was correctly installed and inside it there should already be a working Gun instance. Unless there were any disconnects (not yet sure how to tackle that).

It took quite a few trials and errors to get Gun to work inside the ServiceWorker and there are still some hacks left around... I'll look into the [Service Workers Lifecycle documentation](https://developers.google.com/web/fundamentals/primers/service-workers/lifecycle) to make sure Gun is properly set up (minus the hacks). 

Thanks for chipping in @amark!

**Comment by davide** [comment]
@bugs181 I mislead you in my previous comment!
- the ServiceWorker installation doesn't create the Gun instance right away
- when the ServiceWorker code is first loaded it created a "connect" promise that resolves the Gun instance
- the "connect" promise is what's used inside the "fetch" callback
- the Gun instantiation happens when the ServiceWorker gets its first fetch request

Changed the p2p-fetch logging to better reflect the sequence of events and [removed the hacks](https://github.com/davide/p2p-fetch/commit/e8fc24d6daf03a642905a4b02cb7b0cc18935c3c) I mentioned earlier.
I was relying on the Gun 'hi' event to resolve the "connect" promise but never saw that triggering (any idea why?). That (and frustration) got me to add the dumb 4s second resolve timeout - now dropped with no negative consequences.

New demo URLs:
https://lego-star-wars-fvenmeeizt.now.sh/
https://lego-star-wars-fvenmeeizt.now.sh/?FORCE_P2P=true

The replication steps remain the same, but the first load time is now [4s] faster.

**Comment by davide** [comment]
Well... that was dumb. I was misusing the "connect" promise. Code [updated again](https://github.com/davide/p2p-fetch/commit/02a8f5b2693f2e6ca7eaf213f51c1d7e7a7cf4c0) so that the gun instance is created during:

- Service Worker install; or
- 1st fetch

New demo URLs:
https://lego-star-wars-amofuziquf.now.sh/
https://lego-star-wars-amofuziquf.now.sh/?FORCE_P2P=true

@amark is there a Github issue for @mmalmi 's issue? This might just be a duplicate of that.

**Comment by davide** [comment]
Facepalming like a pro!

The SERVER parameter I had been using was incorrect - hence the lack of 'hi' events from Gun.
This is actually very informative: the issue of .once not triggering happens even when there is NO SERVER CONNECTIVITY.

[Updated the p2p-fetch code and demo](https://github.com/davide/p2p-fetch/commit/5544e506c991aeaa16c8bd12b8492f0751ebd0f6) to actually exercise Gun, giving it 30 seconds to load data from other peers before falling back to the web.

**Comment by davide** [comment]
[Another p2p-fetch update](https://github.com/davide/p2p-fetch/commit/9bd7f6460e080324ed9340fbf05d1396371cee6e), this time to ensure that all page requests are caught -- which [needs some hacks of its own](https://gist.github.com/Rich-Harris/fd6c3c73e6e707e312d7c5d7d0f3b2f9#gistcomment-2737157) to get working.

The current version ([demo link](https://lego-star-wars-qgoetxqmbp.now.sh)) continues with a 30s Gun timeout. This means that the first load (when there's no data) will take those 30s.

**Comment by davide** [comment]
@bugs181 thanks for that last comment! This sentence "assume it's not in Gun and load it immediately" unblocked a partial-solution! :)

We can rely on the synchronous behavior of #once to check if the data is available locally and, if it is, serve it right away. [Got a new commit in with this!](https://github.com/davide/p2p-fetch/commit/31391f093e45d85b4a6abab30219c8bfdb72140a)

With that change things look more deterministic! Here are the new test URLs:
https://lego-star-wars-nlablyatnm.now.sh/
https://lego-star-wars-nlablyatnm.now.sh/?FORCE_P2P=true

The key issue now is that #once

- is not resolving when it has data, instead...
- it resolves only after the configured timeout period has expired (!)

The extreme case is when you open a new browser and pass in ?FORCE_P2P=true (which sets a timeout of 9999999). In that case gun db will just hang in there forever.

While exploring this issue I realized that passing in parameters to a ServiceWorker through it's URL path is actually a VERY BAD IDEA: any changes in the parameters actually setups up *new* ServiceWorkers.
I'll have to change the p2p-fetch API to fix this (using message passing to pass in the ServiceWorker configurations).

**Comment by bugs181** [comment]
> We can rely on the synchronous behavior of #once to check if the data is available locally

Oh. That's even better! Awesome work here. Very exciting stuff.

Instead of using `.once` you may try the `.get(callback)` workaround. It's a higher level function and has a few gotchas - but for realtime makes sense to use it in this case.

**Comment by bugs181** [comment]
@amark it sounds like @mmalmi uncovered something eerily similar to using `.off()` and `.once()` chaining. See https://github.com/amark/gun/issues/685

**Comment by davide** [comment]
@bugs181 thanks again for your help on this issue!

I abandoned the usage of gun.once() with a timeout and went ahead with a [custom "once" implementation](https://github.com/davide/p2p-fetch/commit/72ed86dcb07ff29f5f65119f7a80c86a3bb6b85e) using gun.get(key, callback).

With that change p2p-fetch is now working as expected. See it in action here:
https://lego-star-wars-qjtecfaasn.now.sh?FORCE_P2P=true

**Comment by bugs181** [comment]
Awesome! Yes, sometimes the best course of action is to use the lower level API. I'm glad you were able to come to a solution. Apologies that the API didn't behave as you expected; and that I haven't helped as much as I would of liked. We've all been pretty busy in the Gun community!
